File: dart/lib/src/angle.dart
/* This is free and unencumbered software released into the public domain. */

import 'dart:math' as math;

/// An angle.
///
/// See: https://dogma.dev/Angle
class Angle {
  /// The angle in radians.
  final double radians;

  /// Constructs an angle from radians.
  const Angle(this.radians);

  /// Constructs an angle from radians.
  const Angle.fromRadians(final double radians) : radians = radians;

  /// Constructs an angle from degrees.
  const Angle.fromDegrees(final double degrees)
      : radians = degrees / 180 * math.pi;

  /// Constructs an angle from turns.
  const Angle.fromTurns(final double turns) : radians = turns * 2 * math.pi;

  /// The angle in degrees.
  double get degrees => radians / math.pi * 180;

  /// The angle in turns.
  double get turns => radians / (2 * math.pi);
}

File: js/src/angle.ts
/* This is free and unencumbered software released into the public domain. */

export class Angle {
  private readonly _radians: number;

  public constructor(radians: number) {
    this._radians = radians;
  }

  public static fromDegrees(degrees: number): Angle {
    return new Angle(degrees / 180 * Math.PI);
  }

  public static fromRadians(radians: number): Angle {
    return new Angle(radians);
  }

  public static fromTurns(turns: number): Angle {
    return new Angle(turns * 2 * Math.PI);
  }

  public get degrees(): number {
    return this._radians / Math.PI * 180;
  }

  public get radians(): number {
    return this._radians;
  }

  public get turns(): number {
    return this._radians / (2 * Math.PI);
  }
}
